src/libraries/System.Private.CoreLib/src/System/Random.cs

```diff
@@ -194,6 +196,38 @@ public void GetItems<T>(ReadOnlySpan<T> choices, Span<T> destination)
                 throw new ArgumentException(SR.Arg_EmptySpan, nameof(choices));
             }
 
+            // The most expensive part of this operation is the call to get random data. We can
+            // do so only once rather than per element in a common case:
+            // - The number of choices is <= 256. This let's us get a single byte per choice.
+            // - The number of choices is a power of two. This let's us a byte and simply mask off
+            //   unnecessary bytes cheaply rather than needing to use rejection sampling.
+            // - The type of T doesn't contain any references and is larger than a byte in size.
+            //   This let's us write random bytes into the destination buffer rather than needing
+            //   a scratch buffer. we can use a single RNG.Fill to get all of the required randomness,
+            //   and then walk backwards through the destination, filling in the corresponding choice
+            //   based on the random byte at that position when interpreted as bytes. By walking backwards,
+            //   we won't overrite any random bytes that we still need to read.
+            if (!RuntimeHelpers.IsReferenceOrContainsReferences<T>() && Unsafe.SizeOf<T>() >= 2 &&
+                BitOperations.IsPow2(choices.Length) && choices.Length <= 256)
+            {
+                // Reinterpret the bottom destination.Length bytes of the destination as our
+                // scratch buffer for randomness.
```

Comment:
We could generate a block of random bytes to a stack scratch buffer and fill the output one block at a time.

Reply:
> We could generate a block of random bytes to a stack scratch buffer and fill the output one block at a time.

We can, it just means we either further limit where this can be applied to length stackalloc'able, or we pay the cost of ArrayPool for larger sizes.

Reply:
I meant something like this:
```csharp
Span<byte> randomBytes = stackalloc byte[512];
while (destination.Length > 0)
{
    if (destination.Length < randomBytes.Length)
    {
        randomBytes = randomBytes.Slice(0, destination.Length);
    }

    NextBytes(randomBytes);

    int mask = choices.Length - 1;
    for (int i = 0; i < randomBytes.Length; i++)
    {
        destination[i] = choices[randomBytes[i] & mask];
    }

    destination = destination.Slice(randomBytes.Length);
}
```

Reply:
> I meant something like this:

I see, trading off multiple Fill/NextByte calls for a temporary array.

Reply:
Yeah, when I did the AsBytes for GetString (in a comment in the reporting issue) I thought it was OK because no one other than the thread filling that string knew about the target memory.  In a chat I had with @vcsjones on Friday I said (or was convinced so hard by Kevin that I now recall having said it) that I wouldn't like that for GetItemsCore, but it was OK for GetString.

@jkotas 's approach here is more or less what I was thinking when I paged this out of active thinking on Friday.